lib/sendf.h

```diff
@@ -28,18 +28,33 @@
 
 #include "curl_trc.h"
 
-
-#define CLIENTWRITE_BODY    (1<<0)
-#define CLIENTWRITE_HEADER  (1<<1)
-#define CLIENTWRITE_STATUS  (1<<2) /* the first "header" is the status line */
-#define CLIENTWRITE_CONNECT (1<<3) /* a CONNECT response */
-#define CLIENTWRITE_1XX     (1<<4) /* a 1xx response */
-#define CLIENTWRITE_TRAILER (1<<5) /* a trailer header */
-#define CLIENTWRITE_BOTH   (CLIENTWRITE_BODY|CLIENTWRITE_HEADER)
+/**
+ * Type of data that is being written to the client (application)
+ * - data written can be eiter BODY or META data
+ * - META data is either INFO or HEADER
+ * - INFO is meta information, e.g. not BODY, that cannot be interpreted
+ *   as headers of a response. Example FTP/IMAP pingpong answers.
+ * - HEADER can have additional bits set (more than one)
+ *   - STATUS special "header", e.g. response status line in HTTP
+ *   - CONNECT header was received during proxying the connection
+ *   - 1XX header is part of an intermediate response, e.g. HTTP 1xx code
+ *   - TRAILER header is trailing response data, e.g. HTTP trailers
+ * BODY, INFO and HEADER should not be mixed, as this would lead to
+ * confusion on how to interpret/format/convert the data.
+ */
+#define CLIENTWRITE_BODY    (1<<0) /* non-meta information, BODY */
+#define CLIENTWRITE_INFO    (1<<1) /* meta information, not a HEADER */
```

Comment:
Yeah, that is what I though first as well. But then test cases failed. There are cases, where `data->req.include_header` is set, but pingpong never wrote with `CLIENTWRITE_BODY´. Other protocol handler parts did, though. I believe IMAP is such a case, if memory serves me.

So, before this PR, we had calls to `Curl_client_write()` that:

1. added `CLIENTWRITE_BODY` when `data->req.include_header` was TRUE
2. did not add `CLIENTWRITE_BODY` although `data->req.include_header` was TRUE
3. added `CLIENTWRITE_BODY` irregardless of `data->req.include_header`

Case 1 is now automatically handles in `Curl_client_write()`. Case 2 is now changed to `CLIENTWRITE_INFO`. Case 3 is handled via the special flag set in FTP (and only there because FTP via HTTP PROXY does not want to see the CONNECT headers).

Why do all this? Because with this PR, the flags say what the written data is - and no longer to which callback it shall be passed to. So `CLIENTWRITE_BODY` really is only used for body data.

When this holds true, it is possible to move several things into `Curl_client_write()`:

* content encoding writers, e.g. `data->req.write_stack`
* chunked decoding
* progress updates

making life for transfer loop and protocol handlers easier.

Reply:
To explain the case some more: we have this nice writer stack in `content_encoding.[ch]`. That is also called for `Transfer-Encoding` headers. But `chunked` is not implemented there. Why not?

Well, `chunked` changes the amount of bytes in a response. The chunk framing is not counted against `Content-Length` and download progress updates. In order to count the bytes correctly, it needs to know what of its buffer is *really* the content.

So, when writing received transfer data, one needs to `if()` check for chunked encoding, call it, get back an updated buffer position and length and use that for client writes and progress updates.

If `chunked` were a content_encoding writer, we could just write the received data through the writer stack and it would take care of all this.  Add a "progress" writer *after* "chunked" and updates just work for everyone.